llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,730 @@
+    return;
+  case Value::Number: {
+    auto Num = *Data.getAsNumber();
+    std::ostringstream Oss;
+    Oss << Num;
+    OS << Oss.str();
+    return;
+  }
+  case Value::String: {
+    auto Str = *Data.getAsString();
+    OS << Str.str();
+    return;
+  }
+
+  case Value::Array: {
+    auto Arr = *Data.getAsArray();
+    if (Arr.empty())
+      return;
+    [[fallthrough]];
+  }
+  case Value::Object:
+  case Value::Boolean: {
+    llvm::json::OStream JOS(OS, 2);
+    JOS.value(Data);
+    break;
+  }
+  }
+}
+
+bool isFalsey(const Value &V) {
+  return V.getAsNull() || (V.getAsBoolean() && !V.getAsBoolean().value()) ||
+         (V.getAsArray() && V.getAsArray()->empty()) ||
+         (V.getAsObject() && V.getAsObject()->empty());
+}
+
+void ASTNode::render(const Value &Data, raw_ostream &OS) {
+  ParentContext = &Data;
+  const Value *ContextPtr = T == Root ? ParentContext : findContext();
+  const Value &Context = ContextPtr ? *ContextPtr : nullptr;
+
+  switch (T) {
+  case Root:
+    renderChild(Data, OS);
+    return;
+  case Text:
+    OS << Body;
+    return;
+  case Partial: {
+    auto Partial = Partials->find(Accessor[0]);
+    if (Partial != Partials->end())
+      renderPartial(Data, OS, Partial->getValue());
+    return;
+  }
+  case Variable: {
+    auto Lambda = Lambdas->find(Accessor[0]);
+    if (Lambda != Lambdas->end())
+      renderLambdas(Data, OS, Lambda->getValue());
+    else {
+      EscapeStringStream ES(OS, *Escapes);
+      toMustacheString(Context, ES);
+    }
+    return;
+  }
+  case UnescapeVariable: {
+    auto Lambda = Lambdas->find(Accessor[0]);
+    if (Lambda != Lambdas->end())
+      renderLambdas(Data, OS, Lambda->getValue());
+    else
+      toMustacheString(Context, OS);
+    return;
+  }
+  case Section: {
+    // Sections are not rendered if the context is falsey
+    auto SectionLambda = SectionLambdas->find(Accessor[0]);
+    bool IsLambda = SectionLambda != SectionLambdas->end();
+    if (isFalsey(Context) && !IsLambda)
+      return;
+
+    if (IsLambda) {
+      renderSectionLambdas(Data, OS, SectionLambda->getValue());
+      return;
+    }
+
+    if (Context.getAsArray()) {
+      const json::Array *Arr = Context.getAsArray();
+      for (const Value &V : *Arr)
+        renderChild(V, OS);
+      return;
+    }
+    renderChild(Context, OS);
+    return;
+  }
+  case InvertSection: {
+    bool IsLambda = SectionLambdas->find(Accessor[0]) != SectionLambdas->end();
+    if (!isFalsey(Context) || IsLambda)
+      return;
+    renderChild(Context, OS);
+    return;
+  }
+  }
+  llvm_unreachable("Invalid ASTNode type");
+}
+
+const Value *ASTNode::findContext() {
+  // The mustache spec allows for dot notation to access nested values
+  // a single dot refers to the current context.
+  // We attempt to find the JSON context in the current node, if it is not
+  // found, then we traverse the parent nodes to find the context until we
+  // reach the root node or the context is found
```

Comment:
```suggestion
  // The mustache spec allows for dot notation to access nested values
  // a single dot refers to the current context.
  // We attempt to find the JSON context in the current node, if it is not
  // found, then we traverse the parent nodes to find the context until we
  // reach the root node or the context is found.
```
May also be good to re-wrap the lines, so there isn't a newline between the sentences.